### ISSUE TITLE
Question regarding cls_embedding

### ISSUE THREAD

**Issue by HuangShiqi128** [bytedance/fc-clip#18]
Dear Author,

Thank you for presenting such a nice work.

May I ask why you use the sum of maskpool_embedding and decoder output as class_embed?

Many papers directly use decoder output to derive class_emb, what is the advantage of using maskpool_embedding?

Regards,
Shiqi

**Comment by cornettoyu** [comment]
Hi,

I think this #12 is related and could help answer your question.

Best,

**Comment by HuangShiqi128** [comment]
Hi,

Thank you for your reply.

Yes, I have checked the comments, but it seems that it mainly discussed how and where maskpool_embedding is used in Odise. My doubt is why you use maskpool_embeding as class_emb (in your initial implementation) instead of directly follow the common usage of decoder_output?  I am just curious about the choice and advantages of doing so.

Regards,
Shiqi

**Comment by cornettoyu** [comment]
> Hi,
> 
> Thank you for your reply.
> 
> Yes, I have checked the comments, but it seems that it mainly discussed how and where maskpool_embedding is used in Odise. My doubt is why you use maskpool_embeding as class_emb (in your initial implementation) instead of directly follow the common usage of decoder_output? I am just curious about the choice and advantages of doing so.
> 
> Regards, Shiqi

Thanks for your question. We have not tried using object query for class_embedding (we tried mask pooled decoder output and the sum of both), and we observed that sum = mask pooled decoder output. The (initial) design follows early works such as OpenSeg, which I assume is because the mask pooled decoder output could provide a better initialization for class embeddings (imagine that you have zero-init for the decoder part, the decoder output pixel feature should share a high similarity with CLIP image feature, and mask-pooling over it can give a feature vector that works better with CLIP text embedding). That being said, we have not conducted systematic ablation on this design we are not sure what if using object query as in other close-vocab seg models. Hope it helps.

Best,
